telegram/ext/filters.py

```diff
@@ -850,29 +851,68 @@ class user(BaseFilter):
         Examples:
             ``MessageHandler(Filters.user(1234), callback_method)``
 
+        Attributes:
+            user_ids(set(:obj:`int`), optional): Which user ID(s) to allow through.
+            usernames(set(:obj:`str`), optional): Which username(s) (without leading '@') to allow
+                through.
         Args:
-            user_id(:obj:`int` | List[:obj:`int`], optional): Which user ID(s) to allow through.
-            username(:obj:`str` | List[:obj:`str`], optional): Which username(s) to allow through.
-                If username starts with '@' symbol, it will be ignored.
+            user_id(:obj:`int` | List[:obj:`int`], optional): Which user ID(s) to allow
+                through.
+            username(:obj:`str` | List[:obj:`str`], optional): Which username(s) to allow
+                through. If username starts with '@' symbol, it will be ignored.
 
         Raises:
             ValueError: If chat_id and username are both present, or neither is.
 
         """
 
         def __init__(self, user_id=None, username=None):
-            if not (bool(user_id) ^ bool(username)):
+            if (user_id is None) == (username is None):
                 raise ValueError('One and only one of user_id or username must be used')
-            if user_id is not None and isinstance(user_id, int):
-                self.user_ids = [user_id]
-            else:
-                self.user_ids = user_id
-            if username is None:
-                self.usernames = username
-            elif isinstance(username, string_types):
-                self.usernames = [username.replace('@', '')]
-            else:
-                self.usernames = [user.replace('@', '') for user in username]
+
+            self._user_ids_lock = Lock()
+            self._usernames_lock = Lock()
+
+            # Initialize in a way that will not fail the first setter calls
+            self._user_ids = user_id
+            self._usernames = username
+            # Actually initialize
+            self.user_ids = user_id
+            self.usernames = username
+
+        @property
+        def user_ids(self):
+            with self._user_ids_lock:
+                return self._user_ids
```

Comment:
If you return the (mutable) object itself, it might be changed again after the lock is released. or even the caller might change it. you need to return a copy of the object (preferably frozenset, as it won't give any illusions to the caller).

So:
```suggestion
            with self.__lock:
                return frozenset(self._user_ids)
```

Reply:
Just a thought: This way something like `filter.user_ids.add(chat_id)` won't work and instead one would have to do

```python
current_ids = set(filer.user_ids)
current_ids.add(chat_id)
filters.user_ids = current_ids
```

but that's probably the only really save way to do it?

Reply:
you are correct * 2
in order to be thread safe, we must control the modification of the internal objects ourselves.
also, the fact we're using set is an implementation detail. we shouldn't be exposing it as an API.

Reply:
Fair enough. Resolving.